code/modules/martial_arts/martial.dm

```diff
@@ -428,5 +446,40 @@
 		return ..()
 	return 0
 
+/obj/screen/combo
+	icon_state = ""
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	screen_loc = ui_combo
+	layer = ABOVE_HUD_LAYER
+	var/timerid
+	var/streak
+
+/obj/screen/combo/proc/clear_streak()
+	cut_overlays()
+	streak = ""
+	icon_state = ""
+
+/obj/screen/combo/update_icon(updates, _streak)
+	streak = _streak
+	return ..()
+
+/obj/screen/combo/update_overlays()
+	. = list()
+	for(var/i in 1 to length(streak))
+		var/intent_text = copytext(streak, i, i + 1)
+		var/image/intent_icon = image(icon, src, "combo_[intent_text]")
+		intent_icon.pixel_x = 16 * (i - 1) - 8 * length(streak)
+		. += intent_icon
+
+/obj/screen/combo/update_icon_state()
+	icon_state = ""
+	if(timerid)
+		deltimer(timerid)
+	if(!streak)
+		return
+	timerid = addtimer(CALLBACK(src, PROC_REF(clear_streak)), COMBO_ALIVE_TIME, TIMER_UNIQUE | TIMER_STOPPABLE)
```

Comment:
I'm not keen on having an icon determine when the combo should be cleared. Wouldn't it be better to have this callback on the martial_arts itself? Currently, it is done lazy, but since you need a timer anyways, it can be done using a timer there.